internal/transport/http2_client.go

```diff
@@ -1279,59 +1296,112 @@ func (t *http2Client) reader() {
 	}
 }
 
-// keepalive running in a separate goroutune makes sure the connection is alive by sending pings.
+// activityMonitory reads from the activity channel (which is written to, when
+// there is a read), and updates the lastRead.time atomic.
+func (t *http2Client) activityMonitor() {
+	for {
+		select {
+		case <-t.lr.ch:
+			atomic.StoreInt64(&t.lr.time, time.Now().Unix())
+		case <-t.ctx.Done():
+			return
+		}
+	}
+}
+
+// pinger sends out a PING frame, and waits for one of the following events to
+// happen (kp.Timeout expires, top-level context is done, or the done channel
+// is written to (this usually happens when there is read activity).
+func (t *http2Client) pinger(send bool, done chan struct{}) {
+	if send {
+		if channelz.IsOn() {
+			atomic.AddInt64(&t.czData.kpCount, 1)
+		}
+		t.controlBuf.put(&ping{data: [8]byte{}})
+	}
+
+	timer := time.NewTimer(t.kp.Timeout)
+	select {
+	case <-timer.C:
+		t.Close()
+		return
+	case <-t.ctx.Done():
+		if !timer.Stop() {
+			<-timer.C
+		}
+		return
+	case <-done:
+		if !timer.Stop() {
+			<-timer.C
+		}
+		return
+	}
+}
+
+// keepalive running in a separate goroutune makes sure the connection is alive
+// by sending pings.
 func (t *http2Client) keepalive() {
```

Comment:
Hey @easwars,

I think spawning a new go routine to keep track of the timeout for each ping sent may hurt the performance. Can we use a single go routine to keep track of Time and Timeout at the same time? I have this suggestion:
```golang
func minTime(a, b time.Duration) time.Duration {
	if a < b {
		return a
	}
	return b
}

func (t *http2Client) keepalive() {
	p := &ping{data: [8]byte{}}
	// True iff a PING sent, no data is received since then and didn't timeout.
	outstandingPing := false
	var ackWaitDuration time.Duration
	for {
		dataRead := atomic.CompareAndSwapUint32(&t.activity, 1, 0)
		if dataRead {
			outstandingPing = false
		}

		// If there was an outstanding ping, last timer was a timeout timer and the ping timed out
		if outstandingPing && ackWaitDuration == 0 {
			t.Close()
			return
		}

		// Last timer was a time timer

		// No data read
		if !dataRead {
			t.mu.Lock()
			dormant := len(t.activeStreams) < 1 && !t.kp.PermitWithoutStream
			t.mu.Unlock()
			// If not dormant and no outstanding ping then send a PING
			if !dormant && !outstandingPing {
				t.controlBuf.put(p)
				ackWaitDuration = t.kp.Timeout
				outstandingPing = true
			}
		}

		sleepDuration := t.kp.Time
		// If a PING is sent then we set up a timer for minimum of remaining timeout duration and time duration
		if outstandingPing {
			sleepDuration = minTime(t.kp.Time, ackWaitDuration)
			ackWaitDuration -= sleepDuration
		}
		timer := time.NewTimer(sleepDuration)
		select {
		case <-timer.C:
		case <-t.ctx.Done():
			if !timer.Stop() {
				<-timer.C
			}
			return
		}
	}
}
```

Reply:
A couple of notes on this code:
It uses `t.activity` rather than your last read time. I think using last read time is much better. But that requires some changes in the if conditions (time since last read with kp.Time or kp.Timeout)

It is an oversimplification of the dormancy case. I think it is functionally equivalent. We don't send any PINGs if dormancy conditions are met but the go routine keeps working. I think having this go routine sleep if it is dormant is a good idea.

Reply:
I modified the code to look like this:

```golang
func (t *http2Client) keepalive() {
  p := &ping{data: [8]byte{}}
  // True iff a PING has been sent, and no data has been received since then
  // and the PING hasn't timed out.
  var outstandingPing bool
  // Amount of time remaining before which we should receive an ACK for the
  // last sent PING.
  var timeoutLeft time.Duration
  // UnixNanos recorded before we go block on the timer. This is required to
  // check for read activity since then. 
  var prevNano int64
  for {
    dataRead := false
    if prevNano < atomic.LoadInt64(&t.lr.timeNano) {
      // Read activity since the last time we were here.
      outstandingPing = false
      dataRead = true
    }
  
    // Outstanding PING timed out, we are done.
    if outstandingPing && timeoutLeft <= 0 {
      t.Close()
      return
    }

    if !dataRead {
      // Check if keepalive should go dormant.
      t.mu.Lock()
      if len(t.activeStreams) < 1 && !t.kp.PermitWithoutStream {
        // Make awakenKeepalive writable.
        <-t.awakenKeepalive
        t.mu.Unlock()
        select {
        case <-t.awakenKeepalive:
          // If the control gets here a ping has been sent
          // need to reset the timer with keepalive.Timeout.
          timeoutLeft = t.kp.Timeout
          outstandingPing = true
        case <-t.ctx.Done():
          return
        }
      } else {
        t.mu.Unlock()
        if !outstandingPing {
          if channelz.IsOn() {
            atomic.AddInt64(&t.czData.kpCount, 1)
          }
          t.controlBuf.put(p)
          timeoutLeft = t.kp.Timeout
          outstandingPing = true
        }
      } 
    }
    // Amount of kp.Time remaining should be calculated from the time of the
    // last read activity.
    timeLeft := t.kp.Time
    if dataRead {
      timeLeft = time.Duration(atomic.LoadInt64(&t.lr.timeNano)) + t.kp.Time - time.Duration(time.Now().UTC().UnixNano())
    }
    // If a PING is outstanding, the amount of time to sleep here should be the
    // minimum of timeoutLeft and timeLeft.
    sleepDuration := timeLeft
    if outstandingPing {
      sleepDuration = minTime(timeLeft, timeoutLeft)
      timeoutLeft -= sleepDuration
    } 

    prevNano = time.Now().UTC().UnixNano()
    timer := time.NewTimer(sleepDuration)   
    select {
    case <-timer.C:
    case <-t.ctx.Done():
      if !timer.Stop() {
        <-timer.C
      }
      return
    }
  }
}
```

Existing tests seem to pass. I still need to warm up more to this code though, although I do feel this is better. I will also try to see if I can add more tests to increase my confidence level.

Thanks for the suggestion.